### DISCUSSION TITLE
Attempting the SP Rule Optimization

### DISCUSSION THREAD

**Discussion by drewxa** [amzn/ads-advanced-tools-docs#193] [Q&A]
Hello!

I've attempted to utilize the [schedule-based bid rules API](https://advertising.amazon.com/API/docs/en-us/release-notes/index#schedule-based-bid-rules-now-available-for-sponsored-products), but I'm encountering some questions and issues.

1. I successfully associated the campaign with the rule. However, I'm uncertain about the procedure for disassociating. Could you provide guidance on this?

2. While attempting to search for rules using the provided example from the documentation, the API returns an error:
   > Request must contain filters on (1) optimizationRuleId without any other filters, or (2) a combination of campaignId, ruleCategory, and ruleSubCategory.

   My request was structured as follows:
   ```json
   {
     "optimizationRuleFilter": {
       "ruleCategory": {
         "values": [
           "BID"
         ],
         "filterType": "EXACT_MATCH"
       },
       "ruleSubCategory": {
         "values": [
           "SCHEDULE"
         ],
         "filterType": "EXACT_MATCH"
       }
     }
   }
   ```

3. The API `/sp/campaigns/{campaignId}/optimizationRules` returns a 500 error when I send:
   ```json
   {"optimizationRuleIds": [""]}
   ```

4. The success response contains a unclear message with a different campaign ID. Refer to the image ![image](https://github.com/amzn/ads-advanced-tools-docs/assets/3525417/7c0ba86c-146f-47e7-9ae8-11e94a0ba3ed).

5. Could you guide me on how to delete a rule?

6. Additionally, how can I retrieve information about a rule by its ID? Is there a handler similar to `/sp/rules/optimization/{ruleID}`?

7. Lastly, how can I obtain a list of all optimization rules? As mentioned earlier, the search API does not seem to be functioning correctly.

Thank you for your assistance!

**Comment by ramreal** [comment]
Hello @drewxa - Thank you for you post! I will reach out to the appropriate team and get back to you once I hear back from them.

**Reply by drewxa** [reply]
Hello
Is there any news available?

**Reply by ramreal** [reply]
@drewxa - Please see the answers below:

1. Currently, there is no method of disassociating a rule, we have prioritized this task and request you to recheck our documentation in the near future. Instead, for the time being, you can use an update call to set the rule status to paused, which will disable the rule.
2. As stated in our quoted section, the request must also contain a campaignId for those sets of filters. However it does look like our example is incorrect, we will update this.
3. We have updated the endpoint to no longer have this behavior for this input.
4. This incorrect `campaignID` was a relic of a previous implementation; we have updated the endpoint to reflect this, thanks for pointing this out.
5. Similar to disassociate, there is currently no method of deleting a rule, we have prioritized this task and request you to recheck our documentation in the near future. Instead, for the time being, you can use an update call to set the rule status to paused, which will disable the rule.
6. Currently, there is no way for a direct GET on that path; however, please try a SEARCH call with only an optimizationRuleId filter; it will return the rule and all its details.
7. Similar to 6, there is no current functionality for this; we will consider this a feature request and add to our backlog. Thanks!

**Reply by drewxa** [reply]
I appreciate your response, but there are still lingering concerns.

1. For instance, although I can transition a rule to the paused status, this action will deactivate the rule across **all** campaigns, not just in **a specific one**.

6. In order to retrieve information about a rule and employ your suggested method, I would need to store all the rule Id. This, however, is not always feasible or user-friendly. Additionally, the method outlined in the documentation for searching rules using "ruleCategory" and "ruleSubCategory" does not seem to be viable.

Regrettably, it appears that the API is still in the alpha testing stage. I hope for a swift resolution to all the issues at hand. Your prompt attention to these matters would be highly valued.
